src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs

```diff
@@ -496,20 +504,38 @@ internal bool ReportDiagnostics(IEnumerable<Diagnostic> diagnostics, TextWriter
                     // we previously created.
                     //this assert isn't valid if we change the design to not bail out after each phase.
                     //System.Diagnostics.Debug.Assert(diag.Severity != DiagnosticSeverity.Error);
-                    continue;
+                    return;
                 }
                 else if (diag.Severity == DiagnosticSeverity.Hidden)
                 {
                     // Not reported from the command-line compiler.
-                    continue;
+                    return;
                 }
 
                 // We want to report diagnostics with source suppression in the error log file.
                 // However, these diagnostics should not be reported on the console output.
                 errorLoggerOpt?.LogDiagnostic(diag);
+
+                // If the diagnostic was suppressed by one or more DiagnosticSuppressor(s), then we report info diagnostics for each suppression
+                // so that the suppression information is available in the binary logs and verbose build logs.
+                if (diag.ProgrammaticSuppressionInfo != null)
+                {
+                    foreach (var (id, justification) in diag.ProgrammaticSuppressionInfo.Suppressions)
+                    {
+                        // Diagnostic '{0}' was programmatically suppressed by a DiagnosticSuppressor with suppresion ID '{1}' and justification '{2}'
+                        var suppressionDiag = Diagnostic.Create(SuppressionDiagnosticDescriptor, diag.Location, diag.Id, id, justification);
```

Comment:
>diag.Id [](start = 112, length = 7)

Consider including more information than just an id. For example, in some cases compiler incorporates another DiagnosticInfo as an argument. This way when the message is formatted, the message for the argument is formatted and included as well. This would make the message more informative.

Reply:
Sounds reasonable, I can update it to also contain the original diagnostic message.

Reply:
> Sounds reasonable, I can update it to also contain the original diagnostic message.

We need to make sure it is localized the same way as the rest of the message, i.e. shouldn't probably be localized here. 

---
In reply to: [292201242](https://github.com/dotnet/roslyn/pull/36067#discussion_r292201242) [](ancestors = 292201242)

Reply:
I was thinking of calling [Diagnostic.GetMessage](http://source.roslyn.io/#Microsoft.CodeAnalysis/Diagnostic/Diagnostic.cs,301) API with the [Culture](http://source.roslyn.io/#Microsoft.CodeAnalysis/CommandLine/CommonCompiler.cs,f3865321f1c10bb8,references) used by the command line compiler.

Reply:
> I was thinking of calling Diagnostic.GetMessage API with the Culture used by the command line compiler and embedding it as an additional argument

In general, it is error prone to localize any message arguments before someone requested the message. Ideally, this should be a delayed argument, similar to localizable string, etc. 

---
In reply to: [292217870](https://github.com/dotnet/roslyn/pull/36067#discussion_r292217870) [](ancestors = 292217870)

Reply:
BTW, we don't even want to touch resources and build the message for the underlying diagnostics until someone explicitly asks for that.

---
In reply to: [292565394](https://github.com/dotnet/roslyn/pull/36067#discussion_r292565394) [](ancestors = 292565394,292217870)

Reply:
These special suppression info diagnostics are not returned via any API - we immediately call PrintError on it in this same method below and drop them on the floor. Given this diagnostic is specifically generated for command line compilation scenario where are emitting the message, is it worth optimizing and avoid using GetMessage here?

Reply:
> is it worth optimizing and avoid using GetMessage here?

Why not? The ```PrintError``` is called conditionally. The ```_reportedDiagnostics``` is likely to hold on to the ```suppressionDiag``` and won't let the message to go away even after the ```PrintError``` call.

---
In reply to: [292572828](https://github.com/dotnet/roslyn/pull/36067#discussion_r292572828) [](ancestors = 292572828)

Reply:
Sounds good - I will push a separate commit for this.